bin/deployment/generate_server_certs.sh

```diff
@@ -0,0 +1,231 @@
+                    be self-signed.
+  --cert_out <file> The output filename for the generated CA key; defaults to
+                    \"${CERT_OUT}\"
+  --key_out <file>  The output filename for the generated CA key; defaults to
+                    \"${KEY_OUT}\"
+  --days <days>     The number of days for which the generated cert will be
+                    valid; defaults to ${CERT_DAYS}
+  --bits <bits>     The size of the generated key in bits; defaults to ${KEY_BITS}
+  --hostname <host> An additional hostname to include in the generated cert's
+                    subjectAltName; may be specified multiple times
+  --hostaddr <addr> An additional IP address to include in the generated cert's
+                    subjectAltName; may be specified multiple times
+  -f | --force      Force regeneration and overwrite existing files when the
+                    cert or key already exists; defaults to false
+  -t | --trust      Add the generated cert to the local CA trust; defaults to
+                    false, requires root
+  -h | --help       Display command usage information
+"
+
+while true; do
+    case "$1" in
+        --ca_cert )
+            CA_CERT=$2
+            shift 2 ;;
+
+        --ca_key )
+            CA_KEY=$2
+            shift 2 ;;
+
+        --cert_out )
+            CERT_OUT=$2
+            shift 2 ;;
+
+        --key_out )
+            KEY_OUT=$2
+            shift 2 ;;
+
+        --days )
+            CERT_DAYS=$2
+            shift 2 ;;
+
+        --bits )
+            KEY_BITS=$2
+            shift 2 ;;
+
+        --hostname )
+            SAN+=("DNS:$2")
+            shift 2 ;;
+
+        --hostaddr )
+            SAN+=("IP:$2")
+            shift 2 ;;
+
+        -f | --force )
+            FORCE_REGEN=1
+            shift ;;
+
+        -t | --trust )
+            TRUST=1
+            shift ;;
+
+        -h | --help )
+            echo "${USAGE_TEXT}"
+            exit 0 ;;
+
+        -- ) shift; break ;;
+        * ) if [ ! -z $1 ]; then
+                echo "Invalid option: $1"
+                echo "${USAGE_TEXT}"
+                exit 1
+            fi
+            break ;;
+    esac
+done
+
+fail() {
+    if [ ! -z "$1" ]; then
+        >&2 echo "$1"
+    fi
+
+    exit 1
+}
+
+check_dependency() {
+    if ! command -v $1 > /dev/null; then
+        fail "Required dependency $1 not detected; aborting..."
+    fi
+}
+
+# Fail on any error
+set -e
+
+# Ensure we have all of our dependencies available
+check_dependency openssl
+
+
+
+# If the user has specified that the resultant cert should be trusted, they must also invoke this
+# script as root
+if [ $EUID -ne 0 ] && [ $TRUST -eq 1 ]; then
+    fail "Error: CA trust update must be performed as root"
+fi
+
+# Make sure we're not overwriting existing files by accident
+if [ -f $CERT_OUT ] && [ $FORCE_REGEN -ne 1 ]; then
+    >&2 echo "Warning: Certificate file \"${CERT_OUT}\" already exists; use --force to regenerate"
+
+    exit 0
```

Comment:
I waffled on this a bit myself, and maybe this is a problem of how I'm framing the usage.

The reason I opted to leave it as not an error is due to how the script is used. The call without -f/--force is expected to behave like an idempotent operation that ensures the specific cert exists. Now, this could easily change such that the caller checks if the cert is absent or not and then this would be an error state since it won't overwrite an existing cert by default.

I'm fine with either case here, but if we opt for this to error, the deploy script will need to be updated to trap that error so we don't fail out on the case where the cert already exists.